### PR TITLE
feat(instrumentation-mongoose): Support v7 and v8

### PR DESCRIPTION
**Pull request by onurtemizkan**
Resolves: https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1606

## Which problem is this PR solving?

Related: https://github.com/getsentry/sentry-javascript/issues/11499

- This PR adds support for Mongoose versions [7](https://mongoosejs.com/docs/migrating_to_7.html) and [8](https://mongoosejs.com/docs/migrating_to_8.html) to the instrumentation.

#### Mongoose 7

- The instrumentation already supports async Mongoose operations.
- Moved tests using callbacks into a separate Mongoose v5/v6 test file
- Skipped instrumentation for deprecated `remove` [Mongoose Docs](https://mongoosejs.com/docs/migrating_to_7.html#removed-remove)

#### Mongoose 8

- Skipped instrumentation for deprecated `count` [Mongoose Docs](https://mongoosejs.com/docs/migrating_to_8.html#removed-count)
- Skipped instrumentation for deprecated `findOneAndRemove` [Mongoose Docs](https://mongoosejs.com/docs/migrating_to_8.html#removed-findoneandremove)

The current set of tests is passing without any issues, except `findOneAndUpdate`, which starts to create only one span, starting from version 7. I separated that test for v7/v8.

To avoid separating the package because of type mismatches, I have added a few `// @ts-expect-error`s inside tests with their reasons.

### CODE REVIEW

**Comment by trentm** [comment]
@onurtemizkan I haven't reviewed this PR, yet. A side issue: I notice that you are **force**-pushing. If you are able, please don't force-push to a PR branch, at least not after there have been any reviews on it. It makes it harder to do subsequent re-reviews because it isn't clear what has changed from a previous commit.

**Comment by onurtemizkan** [comment]
Thanks for looking into this @trentm, and yes I'll be careful not to force-push on reviewed code. FWIW these last couple were rebases on main.

File: plugins/node/instrumentation-mongoose/test/mongoose-v5-v6.test.ts [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    assertSpan(spans[0] as ReadableSpan);
+    expect(spans[0].attributes[SEMATTRS_DB_OPERATION]).toBe('count');
+    const statement = getStatement(spans[0] as ReadableSpan);
+    expect(statement.options).toEqual({});
+    expect(statement.condition).toEqual({});
+  });
+
+  it('instrumenting update operation [deprecated]', async () => {
+    // @ts-ignore - v7 removed `update` method
+    // https://mongoosejs.com/docs/migrating_to_7.html#removed-remove
```
Comment (trentm) [inline plugins/node/instrumentation-mongoose/test/mongoose-v5-v6.test.ts:202]:
```suggestion
    // https://mongoosejs.com/docs/migrating_to_7.html#removed-update
```

File: plugins/node/instrumentation-mongoose/.tav.yml [hunk 2]
```diff
@@ -1,5 +1,6 @@
...
-  # Test all the latest minor versions in the range ">=5.9.7 <7".
-  versions: "5.9.7 || 5.9.29 || 5.10.19 || 5.11.20 || 5.12.15 || 5.13.21 || 6.0.15 || 6.1.10 || 6.2.11 || 6.3.9 || 6.4.7 || 6.5.5 || 6.6.7 || 6.7.5 || 6.8.4 || 6.9.3 || 6.10.5 || 6.11.6 || ^6.12.3"
-  commands:
-    - npm run test
+mongoose:
+  - versions: ">=5.9.7 <7"
+    commands: npm run test-v5-v6
+  - versions: ">=7 <9"
+    node: '>=15.0.0'
+    commands: npm run test-v7-v8
```
Comment (trentm) [inline plugins/node/instrumentation-mongoose/.tav.yml:6]:
```suggestion
mongoose:
  - versions:
      include: ">=5.9.7 <7"
      mode: latest-minors
    commands: npm run test-v5-v6
  - versions:
      include: ">=7 <8"
      mode: latest-minors
    node: '>=14.20.1'
    commands: npm run test-v7-v8
  - versions:
      include: ">=8 <9"
      mode: latest-minors
    node: '>=16.20.1'
    commands: npm run test-v7-v8
```

The current state in your PR results in testing almost *300 releases* of mongoose:

```
% npx tav --verbose --dry-run | \rg matching
-- 75 package versions matching filter: 7.0.0, 7.0.1, 7.0.2, 7.0.3, 7.0.4, 7.0.5, 7.1.0, 7.1.1, 7.1.2, 7.2.0, 7.2.1, 7.2.2, 7.2.3, 7.2.4, 7.3.0, 7.3.1, 7.3.2, 7.3.3, 7.3.4, 7.4.0, 7.4.1, 7.4.2, 7.4.3, 7.4.4, 7.4.5, 7.5.0, 7.5.1, 7.5.2, 7.5.3, 7.5.4, 7.6.0, 7.6.1, 7.6.2, 7.6.3, 7.6.4, 8.0.0, 7.6.5, 8.0.1, 7.6.6, 8.0.2, 7.6.7, 8.0.3, 7.6.8, 8.0.4, 8.1.0, 8.1.1, 8.1.2, 8.1.3, 8.2.0, 7.6.9, 8.2.1, 7.6.10, 8.2.2, 8.2.3, 8.2.4, 8.3.0, 8.3.1, 7.6.11, 8.3.2, 8.3.3, 8.3.4, 8.3.5, 8.4.0, 7.6.12, 8.4.1, 7.6.13, 8.4.2, 8.4.3, 7.7.0, 8.4.4, 8.4.5, 8.5.0, 8.5.1, 7.8.0, 8.5.2
-- 212 package versions matching filter: 5.9.7, 5.9.9, 5.9.10, 5.9.11, 5.9.12, 5.9.13, 5.9.14, 5.9.15, 5.9.16, 5.9.17, 5.9.18, 5.9.19, 5.9.20, 5.9.21, 5.9.22, 5.9.23, 5.9.24, 5.9.25, 5.9.26, 5.9.27, 5.9.28, 5.9.29, 5.10.0, 5.10.1, 5.10.2, 5.10.3, 5.10.4, 5.10.5, 5.10.6, 5.10.7, 5.10.8, 5.10.9, 5.10.10, 5.10.11, 5.10.12, 5.10.13, 5.10.14, 5.10.15, 5.10.16, 5.10.17, 5.10.18, 5.10.19, 5.11.0, 5.11.1, 5.11.2, 5.11.3, 5.11.4, 5.11.5, 5.11.6, 5.11.7, 5.11.8, 5.11.9, 5.11.10, 5.11.11, 5.11.12, 5.11.13, 5.11.14, 5.11.15, 5.11.16, 5.11.17, 5.11.18, 5.11.19, 5.11.20, 5.12.0, 5.12.1, 5.12.2, 5.12.3, 5.12.4, 5.12.5, 5.12.6, 5.12.7, 5.12.8, 5.12.9, 5.12.10, 5.12.11, 5.12.12, 5.12.13, 5.12.14, 5.12.15, 5.13.0, 5.13.1, 5.13.2, 5.13.3, 5.13.4, 5.13.5, 5.13.6, 5.13.7, 5.13.8, 6.0.0, 6.0.1, 6.0.2, 6.0.3, 6.0.4, 5.13.9, 6.0.5, 6.0.6, 6.0.7, 6.0.8, 6.0.9, 5.13.10, 6.0.10, 5.13.11, 6.0.11, 5.13.12, 6.0.12, 5.13.13, 6.0.13, 6.0.14, 6.0.15, 6.1.0, 6.1.1, 6.1.2, 6.1.3, 5.13.14, 6.1.4, 6.1.5, 6.1.6, 6.1.7, 6.1.8, 6.1.9, 6.1.10, 6.2.0, 6.2.1, 6.2.2, 6.2.3, 6.2.4, 6.2.5, 6.2.6, 6.2.7, 6.2.8, 6.2.9, 6.2.10, 6.2.11, 6.3.0, 6.3.1, 6.3.2, 6.3.3, 6.3.4, 6.3.5, 6.3.6, 6.3.7, 6.3.8, 6.3.9, 6.4.0, 6.4.1, 6.4.2, 6.4.3, 6.4.4, 6.4.5, 6.4.6, 6.4.7, 6.5.0, 6.5.1, 6.5.2, 5.13.15, 6.5.3, 6.5.4, 6.5.5, 6.6.0, 6.6.1, 6.6.2, 6.6.3, 6.6.4, 6.6.5, 6.6.6, 6.6.7, 6.7.0, 6.7.1, 6.7.2, 6.7.3, 6.7.4, 6.7.5, 6.8.0, 6.8.1, 6.8.2, 6.8.3, 6.8.4, 6.9.0, 6.9.1, 6.9.2, 5.13.16, 6.9.3, 6.10.0, 6.10.1, 6.10.2, 6.10.3, 6.10.4, 5.13.17, 6.10.5, 6.11.0, 6.11.1, 6.11.2, 5.13.18, 5.13.19, 6.11.3, 5.13.20, 6.11.4, 6.11.5, 6.11.6, 6.12.0, 6.12.1, 5.13.21, 6.12.2, 6.12.3, 6.12.4, 5.13.22, 6.12.5, 6.12.6, 6.12.7, 6.12.8, 6.12.9, 6.13.0
```

My suggestion goes back to just testing each latest minor release using test-all-version's (TAV) relatively recent support for [advanced versions specification](https://github.com/watson/test-all-versions?tab=readme-ov-file#advanced-versions-usage). This gets it back down to around 30 versions tested (which perhaps is still more than really necessary):

```
% npx tav --verbose --dry-run | \rg matching
-- 6 package versions matching filter: 8.0.4, 8.1.3, 8.2.4, 8.3.5, 8.4.5, 8.5.2
-- 9 package versions matching filter: 7.0.5, 7.1.2, 7.2.4, 7.3.4, 7.4.5, 7.5.4, 7.6.13, 7.7.0, 7.8.0
-- 19 package versions matching filter: 5.9.29, 5.10.19, 5.11.20, 5.12.15, 5.13.22, 6.0.15, 6.1.10, 6.2.11, 6.3.9, 6.4.7, 6.5.5, 6.6.7, 6.7.5, 6.8.4, 6.9.3, 6.10.5, 6.11.6, 6.12.9, 6.13.0
```

**Review by trentm — CHANGES_REQUESTED** [review]
Thanks for the PR. This looks good, modulo a couple suggestions below.

Please also apply this change to the README.md:

```
diff --git a/plugins/node/instrumentation-mongoose/README.md b/plugins/node/instrumentation-mongoose/README.md
index c5500b14..31a19b4e 100644
--- a/plugins/node/instrumentation-mongoose/README.md
+++ b/plugins/node/instrumentation-mongoose/README.md
@@ -17,7 +17,7 @@ npm install --save @opentelemetry/instrumentation-mongoose

 ## Supported Versions

-- [`mongoose`](https://www.npmjs.com/package/mongoose) versions `>=5.9.7 <7`
+- [`mongoose`](https://www.npmjs.com/package/mongoose) versions `>=5.9.7 <9`

 ## Usage

```

**Review by trentm — APPROVED** [review]

**Pull request merged**